packages/astro/src/content/utils.ts

```diff
@@ -358,3 +387,59 @@ function search(fs: typeof fsMod, srcDir: URL) {
 	}
 	return { exists: false, url: paths[0] };
 }
+
+export async function getStringifiedLookupMap({
+	contentPaths,
+	contentEntryConfigByExt,
+	root,
+	fs,
+}: {
+	contentEntryConfigByExt: ReturnType<typeof getContentEntryConfigByExtMap>;
+	contentPaths: Pick<ContentPaths, 'contentDir' | 'cacheDir'>;
+	root: URL;
+	fs: typeof fsMod;
+}) {
+	const { contentDir } = contentPaths;
+	const globOpts: FastGlobOptions = {
+		absolute: true,
+		cwd: fileURLToPath(root),
+		fs: {
+			readdir: fs.readdir.bind(fs),
+			readdirSync: fs.readdirSync.bind(fs),
+		},
+	};
+
+	const relContentDir = rootRelativePath(root, contentDir, false);
+	const contentGlob = await glob(
+		`${relContentDir}**/*${getExtGlob([...contentEntryConfigByExt.keys()])}`,
+		globOpts
+	);
+	let filePathByLookupId: {
+		[collection: string]: Record<string, string>;
+	} = {};
+
+	for (const filePath of contentGlob) {
+		const info = getEntryInfo({ contentDir, entry: filePath });
+		if (info instanceof NoCollectionError) continue;
+		const contentEntryConfig = contentEntryConfigByExt.get(extname(filePath));
+		if (!contentEntryConfig) continue;
+
+		const { id, collection, slug: generatedSlug } = info;
+		filePathByLookupId[collection] ??= {};
+		const { slug: frontmatterSlug } = await contentEntryConfig.getEntryInfo({
+			fileUrl: pathToFileURL(filePath),
+			contents: await fs.promises.readFile(filePath, 'utf-8'),
+		});
+		const slug = getEntrySlug({ id, collection, generatedSlug, frontmatterSlug });
+		filePathByLookupId[collection][slug] = rootRelativePath(root, filePath);
+	}
```

Comment:
IIUC this is where we move most of the runtime work to build time. Since this does an `await` for each file, should we `.map` and use a `Promise.all`?

Reply:
We're usually wary of `Promise.all` for memory usage. But since this is a one-time cost, I'll give it a try. Also curious if a `frontmatterSlug` cache is worthwhile for dev server runtime performance